### PR TITLE
Axl/s3 multipart uploads

### LINKED ISSUE #2494: S3 publishing fails for artifacts larger than 5GiB
When publishing artifacts to an S3-backed repository, the upload will fail if an individual artifact is larger than 5GiB.

The cause is that AWS S3 limits the file size to 5GiB when using the simple put object API, currently used in S3Client. The solution is to instead use the multi-part API, to transparently enable support for up to 5TiB files.

I've offered to implement multi-part support in this thread: https://groups.google.com/forum/#!topic/gradle-dev/auNi3fpg-f8

### PR DESCRIPTION
**Pull request by judgeaxl**
Fixes #2494 

### Context
When uploading larger files, e.g. >50MiB, it's beneficial to utilize the S3 multi-part upload API over a single stream, to increase speed. The implementation will automatically switch to use the multi-part API if the upload is larger than 50MiB. I'm not familiar enough with the Gradle internals to know where this would best be configured by a user, so I expect a discussion about that to follow.

I'm not sure how many people upload large publications, but for those who do, utilizing the multi-part upload API should give speed improvements, as well as support for files larger than 5GiB.

There was an email thread but I can no longer locate it. It was from 2017, and is referenced in the issue above. This work fell off my radar at the time, but got resurrected recently in a new project.

### Contributor Checklist
- [X] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [X] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [X] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [X] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [X] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [X] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [X] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by judgeaxl** [comment]
I found the old discussion here: https://groups.google.com/forum/#!topic/gradle-dev/auNi3fpg-f8

**Comment by judgeaxl** [comment]
Since I still would like this to get accepted at some point, what's the best way to proceed? Shall I rebase against master and make sure all the checks pass, or keep this old branch? I.e are the PR checks running against an attempted merge, or against the branch?

**Comment by judgeaxl** [comment]
I still want this merged if possible. Since it's quite far behind master, should I rebase? The contribution guide doesn't really mention the workflow for old PRs.

I also don't have access to check the failing Team City builds, so I have no idea what to do about those.

**Comment by JLLeitschuh** [comment]
@Mergifyio rebase

**Comment by JLLeitschuh** [comment]
@bot-gradle test this please

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java [hunk 1]
```diff
@@ -132,6 +147,48 @@ public void put(InputStream inputStream, Long contentLength, URI destination) {
...
+    private void putMultiPartObject(InputStream inputStream, Long contentLength, URI destination) {
+        try {
+            S3RegionalResource s3RegionalResource = new S3RegionalResource(destination);
+            String bucketName = s3RegionalResource.getBucketName();
+            String s3BucketKey = s3RegionalResource.getKey();
+            configureClient(s3RegionalResource);
+            List<PartETag> partETags = new ArrayList<>();
+            InitiateMultipartUploadRequest initRequest = new InitiateMultipartUploadRequest(bucketName, s3BucketKey)
+                .withCannedACL(CannedAccessControlList.BucketOwnerFullControl);
+            InitiateMultipartUploadResult initResponse = amazonS3Client.initiateMultipartUpload(initRequest);
```
Comment (JLLeitschuh) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java:159]: Are any of these `Closable` and/or need to be closed?

Comment (judgeaxl) [reply subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java:159]: I haven't found anything in the documentation or examples that suggest one needs to close or dispose of anything other than via the abortMultipartUpload() method in the exception handler.

**Review by JLLeitschuh — APPROVED** [review]
LGTM! Thanks for your contribution! Much appreciated!

**Pull request merged**